membrane_curvature/lib/mods.py

```diff
@@ -24,66 +38,58 @@ def grid_map(coords, factor):
     return index_grid_l, index_grid_m
 
 
-def core_fast_leaflet(universe, z_Ref, n_cells, selection, max_width):
+def derive_surface(n_cells, selection, max_width):
```

Comment:
Hi @ojeda-e I'm looking forward to being able to use your membrane curvature tool! I'm not sure if this fits your purpose exactly, but you could use [`scipy.stats.binned_statistic_2d`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.binned_statistic_2d.html) to generate a surface of all your beads in one go.

Reply:
> Hi @ojeda-e, I think I had something different in mind when I requested that you only use numpy etc and break it down to basics. Could you have a go at writing the function with the input arguments I suggested? I haven't had a look at tests because I assume they'll change with the new function.
> 
> @p-j-smith had a good suggestion with the scipy function. When I said "numpy only" that was probably too restrictive, sorry; totally on me. For future reference, you should consider any package safe if it's something that MDAnalysis considers a ["core dependency"](https://userguide.mdanalysis.org/stable/module_imports.html#list-of-core-module-dependencies). I actually find it preferable to use numpy/scipy functions because they're usually written by people much smarter than me, who've thought about the problem longer than me -- this usually results in faster, better code.
> 
> If you try something like the code below, you might make this a very short function!
> 
> ```python
> x, y, z = coordinates.T
> scipy.stats.binned_statistic_2d(x, y, z, statistic="mean", bins=..., range=...)
> ```

Hi @lilyminium, the PR has a lot of different feedback at this point and I ended up following the suggestion from @richardjgowers which conflicts with this one. Could we defer this to an issue as it would require me to add a new dependency and can be done in parallel with other work?

Reply:
> def get_z_surface(coordinates, n_x_bins=10, n_y_bins=10, x_range=(0, 100), y_range=(0, 100)):

I don't think the feedback about the function signature contradicts @richardjgowers. You don't have to use scipy now if you don't want to -- the main gist is to have the main function work solely on numpy arrays of positions. I think you'll find this helpful for refactoring in #41; if you have a look at existing AnalysisBase classes, the important functions very rarely work on MDAnalysis Universes. Instead, the positions are extracted in `_single_frame` and these numpy arrays are used for the actual computation.

You've already done most of the work, to be honest. The simplest, most immediate conversion would be:

```python
def derive_surface(n_cells, selection, max_width):
	coordinates = selection.positions
	return get_z_surface(coordinates, n_x_bins=n_cells, n_y_bins=n_cells,
					     x_range=(0, max_width), y_range=(0, max_width))

def get_z_surface(coordinates, n_x_bins=10, n_y_bins=10, x_range=(0, 100), y_range=(0, 100)):
	z_ref = np.zeros((n_x_bins, n_y_bins))
	grid_z_coordinates = np.zeros((n_x_bins, n_y_bins))
    grid_norm_unit = np.zeros((n_x_bins, n_y_bins))

    x_factor = n_x_bins / (x_range[1] - x_range[0])
    y_factor = n_y_bins / (y_range[1] - y_range[0])
    x_coords, y_coords, z_coords = coordinates.T
    cell_x_floor = np.floor(x_coords * x_factor).astype(int)
    cell_y_floor = np.floor(y_coords * y_factor).astype(int)
    for l, m, z in zip(cell_x_floor, cell_y_floor, z_coords):
    	... # rest of current derive_surface
```

Reply:
I also wouldn't be afraid of adding a new dependency (especially from the MDAnalysis core dependencies). It's mostly a matter of adding the package name to:
* https://github.com/MDAnalysis/membrane-curvature/blob/main/devtools/conda-envs/test_env.yaml
* https://github.com/MDAnalysis/membrane-curvature/blob/main/docs/requirements.yaml

Reply:
Thanks, I just thought this was going to be a separate issue. In the function you suggest the number of x and y bins is different, which I thought was going to be addressed later in #35 (together with PBC issue #36). 

Would you please confirm that is better to add this change here? @lilyminium 
Another option is to use your suggested function but without independent `x` and `y` (i.e. square arrays) and then complete the work when I address #35 later on?

Reply:
IMO since you know that you want to address #35 in the future, there's no harm in designing your code towards that direction now. If you think it's easier to write and test a function that only takes `get_z_surface(coordinates, n_bins=10, range=(0, 100))` and then modify all the tests you write here for the `get_z_surface(coordinates, n_x_bins=10, n_y_bins=10, x_range=(0, 100), y_range=(0, 100))` signature later, go for it -- but to me it sounds like extra work.

If you have no intention of testing non-square arrays in this PR, you could just pass the same number into `n_x_bins` and `n_y_bins` as shown in the `derive_surface` example above. Then when you want to add rectangular functionality, you don't need to modify the existing tests from this PR.

Reply:
Thanks for the quick reply. I'll add it here then.